llvm_passes/HipPrintf.cpp

```diff
@@ -0,0 +1,203 @@
+public:
+  static char ID;
+  HipPrintfToOpenCLPrintfLegacyPass() : ModulePass(ID) {}
+
+  bool runOnModule(Module &M) override {
+    return false;
+  }
+
+  StringRef getPassName() const override {
+    return "Convert printf calls to OpenCL compatible ones.";
+  }
+};
+
+char HipPrintfToOpenCLPrintfLegacyPass::ID = 0;
+static RegisterPass<HipPrintfToOpenCLPrintfLegacyPass>
+    X("hip-printf", "Convert printf calls to OpenCL compatible ones.");
+
+// Converts the address space of the format string to OpenCL compatible one
+// by creating a copy of it in the module global scope.
+//
+// Also counts the number of format args for replacing the return
+// value of the printf() call with it for CUDA emulation.
+Value* convertFormatString(Value *HipFmtStrArg, Instruction *Before,
+                           unsigned &NumberOfFormatSpecs) {
+
+  Module *M = Before->getParent()->getParent()->getParent();
+
+  Type *Int8Ty = IntegerType::get(M->getContext(), 8);
+  ConstantExpr *CE = cast<ConstantExpr>(HipFmtStrArg);
+  assert(CE->isGEPWithNoNotionalOverIndexing());
+
+  // TODO: Cannot we really get access to the ptr operand in a simpler way
+  // without a new object to delete? Perhaps use the cexpr operand replace API?
+  GetElementPtrInst *GEP =
+    cast<GetElementPtrInst>(CE->getAsInstruction());
+  GlobalVariable *OrigFmtStr =
+    cast<GlobalVariable>(GEP->getPointerOperand());
+
+  Constant *FmtStrData = OrigFmtStr->getInitializer();
+
+  NumberOfFormatSpecs = 0;
+  int I = 0;
+  while (Constant *Chr = FmtStrData->getAggregateElement(I)) {
+    char C = cast<ConstantInt>(Chr)->getZExtValue();
+
+    char NextC = 0;
+    if (Constant *NextChr = FmtStrData->getAggregateElement(I + 1))
+      NextC = cast<ConstantInt>(NextChr)->getZExtValue();
+
+    if (C == '%') {
+      if (NextC == '%') {
+        I += 2;
+      } else {
+        ++NumberOfFormatSpecs;
+        ++I;
+      }
+    } else
+      ++I;
+  }
+
+  GlobalVariable *NewFmtStr = new GlobalVariable(
+      *M, OrigFmtStr->getValueType(), true, OrigFmtStr->getLinkage(),
+      FmtStrData, OrigFmtStr->getName() + ".cl",
+      (GlobalVariable *)nullptr, OrigFmtStr->getThreadLocalMode(),
+      SPIRV_OPENCL_PRINTF_FMT_ARG_AS);
+  NewFmtStr->copyAttributesFrom(OrigFmtStr);
+
+  delete GEP;
+
+  PointerType *OCLPrintfFmtArgT =
+    PointerType::get(Int8Ty, SPIRV_OPENCL_PRINTF_FMT_ARG_AS);
+
+  IntegerType *Int64Ty = Type::getInt64Ty(M->getContext());
+  ConstantInt *Zero = ConstantInt::get(Int64Ty, 0);
+
+  std::array<Constant*, 2> Indices = {Zero, Zero};
+
+  Constant *NewCE =
+    llvm::ConstantExpr::getGetElementPtr(nullptr, NewFmtStr, Indices);
+  return NewCE;
+}
+
+
+PreservedAnalyses HipPrintfToOpenCLPrintfPass::run(
+  Function &F,
+  FunctionAnalysisManager &AM) {
+
+  Module *M = F.getParent();
+  GlobalValue *Printf = M->getNamedValue("printf");
+  GlobalValue *HipPrintf = M->getNamedValue("_hip_printf");
+
+  if (Printf == nullptr) {
```

Comment:
If there happens to be a printf declaration you could also check it does not have any users (hasNUses(0)).

Reply:
It will have uses if there are multiple functions in the module and have converted before. But since you spotted that it's not allowed to push new globals in function passes, this has to be reorganized to the module pass init which should be cleaner.